xds/internal/xdsclient/client.go

```diff
@@ -36,6 +36,20 @@ type XDSClient interface {
 	WatchEndpoints(string, func(xdsresource.EndpointsUpdate, error)) func()
 	ReportLoad(*bootstrap.ServerConfig) (*load.Store, func())
 
+	// WatchResource uses xDS to discover the resource associated with the
```

Comment:
I initially had the `WatchResource` method in here, but at the last minute ended up making an interface on the `xdsresource` package (because I could not use this `XDSClient` interface in there for cyclic dependency reasons). I've moved back to having the method in here.

The ideal state of affairs should be something like this:
- We have an `XDSClient` interface in the `xdsclient` package which contains a `WatchResource` method
- The resource type specific watch functions exposed by the `xdsresource` package will need to accept as its first argument an interface which contains the `WatchResource` method. And callers would pass in the concrete `xdsClient` instance that they have as the first argument. But I couldn't come up with a name for this interface which has a single method `WatchResource`. I couldn't name it `ResourceWatcher`, because we have other interfaces which are named similarly like `GenericResourceWatcher`, `ListenerResourceWatcher` and these interfaces are watchers from the point of view of the `xdsClient` because these contain callbacks for events on the resources being watched. But the interface with the single `WatchResource` method is the one which is actually watching the resource on the management server by sending xDS requests. So, I gave up and named this interface also `XDSClient` in the `xdsresource` pacakge.

Do you have any name suggestions?

Reply:
> because I could not use this XDSClient interface in there for cyclic dependency reasons

Could a third package help here?

Maybe `GenericResourceWatcher` should be `ResourceWatcher`, and `WatchResource` should actually be called `Subscribe`?

Also why `Watcher` and not `Listener`?  We should maybe be consistent with these.  I used `Listener` with ORCA (because Java did).  Then the method to attach the listener would be `XDSClient.Listen` and you'd pass it an `xdsclient/xdsresource.Listener`.  But maybe this is too nit picky.

Reply:
I like `Subscribe` over `WatchResource` even though the latter is the one which is used in C++ (I haven't checked Java). Even with `Subscribe`, the name of the interface would be tricky. We can't name that interface `Subscriber`, because technically subscribers should be the one who are calling `Subscribe`. Maybe I could use something like a `Provider`. In that case, I could have the following packages:

```golang
package xdsresource

type Provider interface {
  Subscribe(rType Type, resourceName string, watcher GenericResourceWatcher) (cancel func())
}
```

```golang
package xdsclient

type XDSClient interface {
  xdsresource.Provider
  ...
}
```

```golang
package xdsresource/listener

// Contains actual resource type specific functionality like implementation of
// the `Decode` method etc.

func Subscribe(p xdsresource.Provider, resourceName string, watcher ResourceWatcher) (cancel func()) {...}
```

What do you think about this approach?

The other nice thing about this approach would be that when someone wants to add a new resource type, they would put that in a brand new package under `xdsresource`.

Reply:
> Also why `Watcher` and not `Listener`?  We should maybe be consistent with these. 

`Listener` is a resource type here and `Watcher` has always been the terminology used for the entity watching a resource. I think changing that would be quite problematic (not just from implementation perspective, but also interop of terms across languages).

Reply:
> We can't name that interface Subscriber, because technically subscribers should be the one who are calling Subscribe

AAAAAAAAHHH I hate this bit of Go idiom-syncracy.  It came up with wrapped errors IIRC, and they eventually punted and did not define the interface at all.

We can't define the _whole_ `XDSClient` interface inside `xdsresource`?  If we can, we can use a type alias to make it still convenient:

```go
package xdsclient

type XDSClient = xdsresource.XDSClient
```

Then we don't need an interface for this one function either.  And `WatchResource` would be fine as a name, too.  (IIUC the big problem with `WatchResource` is needing the interface to be named `ResourceWatcher` which is weird.)